### PR TITLE
[develop] Performance: Add memcached support/Improve database cache

### LINKED ISSUE #4518: Performance: Add Memcached support
See https://friendica.mrpetovan.com/display/735a2029145a96e3a87f168339059262, although this issue is only remotely linked with the initial discussion, I got severely carried away.

### PR DESCRIPTION
**Pull request by MrPetovan**
Closes #4518

Kids, don't do this at home.

At heart, it was only about removing the current awful cache clearing mechanism (see https://friendica.mrpetovan.com/display/735a2029145a96e3a87f168339059262).

And then I realized that we only supported Memcache, but not Memcached for caching. So I created an Interface for cache drivers, but then I realized the Session was using the obsolete Memcache mechanism, so I created a new cache session handler, but then I realized that the config values were undocumented, so I created new ones.

It ended up being a 4-in-1 monster PR, and I apologize in advance.

But look at how good the new code looks! And no more random cache clearing on page load!

Removed config values:
- system.memcache
- system.disable_database_session (undocumented)

Added config values:
- system.cache_driver: **database**, memcache, memcached
- system.session_handler: **database**, cache, native
- system.memcached_host
- system.memcached_port

### CODE REVIEW

**Comment by annando** [comment]
Are you currently running on this branch? Since this touches many basic functions, this would be very good.

File: doc/htconfig.md [hunk 1]
```diff
@@ -58,9 +58,12 @@ Example: To set the automatic database cleanup process add this line to your .ht
 * **max_processes_backend** - Maximum number of concurrent database processes for background tasks. Default value is 5.
 * **max_processes_frontend** - Maximum number of concurrent database processes for foreground tasks. Default value is 20.
 * **min_poll_interval** - minimal distance in minutes between two polls for a contact. Default is 1. Reasonable values are between 1 and 59.
-* **memcache** (Boolean) - Use memcache. To use memcache the PECL extension "memcache" has to be installed and activated.
-* **memcache_host** - Hostname of the memcache daemon. Default is '127.0.0.1'.
-* **memcache_port** - Portnumber of the memcache daemon. Default is 11211.
+* **session_handler** ([database]|cache|native) - Whether to use Cache to store session data or to use PHP native session storage
```
Comment (tobiasd) [inline doc/htconfig.md]: do the square bracket indicate a default value?

File: friendica_test_data.sql [hunk 2]
```diff
@@ -112,13 +112,13 @@ DROP TABLE IF EXISTS `cache`;
 /*!40101 SET @saved_cs_client     = @@character_set_client */;
 /*!40101 SET character_set_client = utf8 */;
 CREATE TABLE `cache` (
```
Comment (tobiasd) [inline friendica_test_data.sql]: have you tested the VM running with that new DB?

File: database.sql [hunk 3]
```diff
@@ -55,13 +55,13 @@ CREATE TABLE IF NOT EXISTS `auth_codes` (
...
-	`updated` datetime NOT NULL DEFAULT '0001-01-01 00:00:00' COMMENT '',
-	 PRIMARY KEY(`k`),
-	 INDEX `expire_mode_updated` (`expire_mode`,`updated`)
-) DEFAULT COLLATE utf8mb4_general_ci;
+	`k` varbinary(255) NOT NULL COMMENT 'cache key',
+	`v` mediumtext COMMENT 'cached serialized value',
+	`expires` datetime NOT NULL DEFAULT '0001-01-01 00:00:00' COMMENT 'datetime of cache expiration',
+	`updated` datetime NOT NULL DEFAULT '0001-01-01 00:00:00' COMMENT 'datetime of cache insertion',
+	PRIMARY KEY (`k`),
+	KEY `k_expires` (`k`,`expires`) USING BTREE
```
Comment (annando) [inline database.sql]: Just some side note: Have you created this file by hand?

File: doc/htconfig.md (hunk 1 above)
Comment (MrPetovan) [reply doc/htconfig.md]: Indeed, I wanted to use bold but it already is used by the setting name.

File: friendica_test_data.sql (hunk 2 above)
Comment (MrPetovan) [reply friendica_test_data.sql]: No, I don't have Vagrant on Windows and I'm not sure how to get it.

File: database.sql (hunk 3 above)
Comment (MrPetovan) [reply database.sql]: No, it's phpMyAdmin output, why?

**Comment by tobiasd** [comment]
> Indeed, I wanted to use bold but it already is used by the setting
> name.

For other settings the default values are written out at the end of the
description of the setting. I think it should be one or the other and
easily readable for those seeking knowledge.

File: database.sql (hunk 3 above)
Comment (annando) [reply database.sql]: Please create it by calling ```scripts/dbstructure.php > database.sql```. This is unified - and easier to perform.

Comment (MrPetovan) [reply database.sql]: Ah, I forgot about that, will do tonight.

**Comment by MrPetovan** [comment]
Ok, I'll change it tonight.

**Comment by MrPetovan** [comment]
Retracted to allow configuring multiple Memcached servers.

**Pull request closed without merging**